README
WMBlueCPU
=========

WMBlueCPU is a dockapp. You'll never guess what it does.
It can do SMP too.

Right-click for a menu.


ChangeLog
Version		Description
###########################
0.9	Added menu and probably some other stuff
0.8     [Insert l33t phraze here]
0.7     I forgot.
0.6	Completely rewritten. Kept the FreeBSD cpu code.
0.5	Thanks to Alexey Dokuchaev <danfe@regency.nsu.ru>,
          now FreeBSD is supported!
	Merged all pixmaps into one, 'pixmap.xpm'
0.4	Now windowed mode will have no decorations,
	 the window can be moved with the mouse.
	Moved pixmaps to xpm/
0.3	Cleaned up options
	Cleaned up the whole thing (but just a little), 'twas a mess
	Every expose event now just calls update_window, not draw_window, so
	  there's less cpu usage here.
0.2	Removed -bw, now not needed :)
	When run with -ns, a panel is created.
	Renamed timer.c and timer.h to timing.c and timing.h
	made -s and -m into one: -s <MILISECS>
0.1	Added the following options:
	-bw - make it run in bad windowmanagers
	-w - make it run in a window
	-ns - don't use the XShape extension


cpu.h
/*
 *  cpu_linux.h
 *
 *  Copyright (C) 2003 Draghicioiu Mihai <misuceldestept@go.ro>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Street #330, Boston, MA 02111-1307, USA.
 */

#ifndef __CPU_H__
#define __CPU_H__

extern long cpu_used;
extern long cpu_total;
extern int  cpu_num;

void cpu_init(int);
void cpu_getusage();

#endif /* __CPU_H__ */


cpu_linux.c
/*
 *  cpu_linux.c - get cpu usage
 *
 *  Copyright (C) 2003 Draghicioiu Mihai <misuceldestept@go.ro>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Street #330, Boston, MA 02111-1307, USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

long cpu_used;
long cpu_total;
int  cpu_num;
long oldused;
long oldtotal;

void cpu_init(int c)
{
 cpu_num = c;
}

void cpu_getusage()
{
 FILE *file;
 long cpu, nice, system, idle, used, total;
 char buf[128];
 int c = -1;

 file = fopen("/proc/stat", "r");
 if(!file)
 {
  perror("/proc/stat");
  exit(1);
 }
 while(c != cpu_num)
 {
  if(fgets(buf, 128, file) == NULL)
  {
   fprintf(stderr, "Cannot stat cpu.\n");
   exit(1);
  }
  sscanf(buf, "%*03s%d %ld %ld %ld %ld", &c, &cpu, &nice, &system, &idle);
 }
 fclose(file);
 used =  cpu + nice + system;
 total = used + idle;
 cpu_used = used - oldused;
 cpu_total = total - oldtotal;
 oldused = used;
 oldtotal = total;
}


cpu_freebsd.c
/*
 *  cpu_freebsd.c - get cpu usage
 *
 *  Copyright (C) 2003 Alexey Dokuchaev <danfe@regency.nsu.ru>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Street #330, Boston, MA 02111-1307, USA.
 */

#include <osreldate.h>
#if __FreeBSD_version > 500100
#include <sys/time.h>
#include <sys/resource.h>
#else
#include <sys/dkstat.h>
#endif
#include <sys/types.h>
#include <fcntl.h>
#include <kvm.h>
#include <nlist.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

long cpu_used;
int  cpu_num;
long oldused;
long oldtotal;
static kvm_t *kd;
static struct nlist nlst[] = { {"_cp_time"}, {0} };

void
cpu_init(int c)
{
	if (!(kd = kvm_open(NULL, NULL, NULL, O_RDONLY, "kvm_open")))
	{
		perror("kvm_open");
		exit(1);
	}

	kvm_nlist(kd, nlst);

	if (!nlst[0].n_type)
	{
		perror("kvm_nlist");
		exit(1);
	}

	seteuid(getuid());
	setegid(getgid());

	if (geteuid() != getuid() || getegid() != getgid())
	{
		perror("sete?id");
		exit(1);
	}
	cpu_num = 0;
}

void cpu_getusage()
{
	long cpu, nice, system, idle, used, total;
	long cpu_time[CPUSTATES];

	if (kvm_read(kd, nlst[0].n_value, &cpu_time, sizeof(cpu_time))
		!= sizeof(cpu_time))
	{
		perror("kvm_read");
		exit(1);
	}

	cpu = cpu_time[CP_USER];
	nice = cpu_time[CP_NICE];
	system = cpu_time[CP_SYS];
	idle = cpu_time[CP_IDLE];

	used = cpu + nice + system;
	total = used + idle;

	cpu_used = used - oldused;
	cpu_total = total - oldtotal;

	oldused = used;
	oldtotal = total;
}


main.c
/*
 *  WMBlueCPU - a cpu monitor
 *
 *  Copyright (C) 2003 Draghicioiu Mihai Andrei <misuceldestept@go.ro>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <sys/select.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/keysym.h>
#include <X11/xpm.h>
#include <X11/extensions/shape.h>

#include "cpu.h"
#include "mwm.h"
#include "menu.h"

#include "pixmap.xpm"
#include "icon.xpm"

#define WINDOW_WIDTH 64
#define WINDOW_HEIGHT 64
#define WINDOW_NAME "WMBlueCPU"
#define ICON_NAME "wmbluecpu"
#define CLASS_NAME "wmbluecpu"
#define CLASS_CLASS "WMBlueCPU"
#define GRAPH_HEIGHT 12
#define GRAPH_LENGTH 18
#define GRAPH_X 6
#define GRAPH_Y 24
#define GRAPH_PT_W 3
#define GRAPH_PT_H 3
#define PERCENT_X 28
#define PERCENT_Y 16
#define CPUNUM_X 13
#define CPUNUM_Y 16

#define OPT_DISPLAY NULL
#define OPT_MILISECS 1000
#define OPT_BGCOLOR  "rgb:00/00/00"
#define OPT_ONCOLOR  "rgb:87/D7/FF"
#define OPT_OFFCOLOR "rgb:00/95/E0"
#define OPT_WINDOW 0
#define OPT_SHAPE 1
#define OPT_CPUNUM 0

int            argc;
char         **argv;
Display       *display;
int            screen;
Colormap       colormap;
Visual        *visual;
int            depth;
GC             backgc, offgc, ongc;
Window         rootwindow, window, iconwindow, mapwindow;
XEvent         event;
int            exitloop;
Atom           wm_delete_window;
Atom           _motif_wm_hints;
MWMHints       mwmhints;
Pixmap         buffer, pixmap, pixmask, iconpixmap;
unsigned long  bgcolor, offcolor, oncolor;
struct timeval tv = { 0, 0 };
int            oldx, oldy;
menu_t *       m;


char *opt_display  = OPT_DISPLAY;
int   opt_milisecs = OPT_MILISECS;
char *opt_bgcolor  = OPT_BGCOLOR;
char *opt_offcolor = OPT_OFFCOLOR;
char *opt_oncolor  = OPT_ONCOLOR;
int   opt_window   = OPT_WINDOW;
int   opt_shape    = OPT_SHAPE;
int   opt_cpunum   = OPT_CPUNUM;

unsigned long get_color(char *colorname)
{
 XColor c, c2;
 c2.pixel = 0;
 XAllocNamedColor(display, colormap, colorname, &c, &c2);
 return c.pixel;
}

void bad_option(int arg)
{
 fprintf(stderr, "%s needs an argument.\n"
                 "Try `%s --help' for more information.\n",
                 argv[arg-1], argv[0]);
 exit(1);
}

void print_usage()
{
 printf("Usage: %s [option] [-c <cpunumber>]\n"
	"Options:\n"
	"    -h,  --help           Print out this help and exit\n"
	"    -v,  --version        Print out version number and exit\n"
	"    -d,  --display  <dpy> The X11 display to connect to\n"
	"    -m,  --milisecs <ms>  The number of milisecs between updates\n"
	"    -b,  --bgcolor  <col> The background color\n"
	"    -f,  --offcolor <col> Color for Off leds\n"
	"    -o,  --oncolor  <col> Color for On leds\n"
	"    -w,  --window         Run in a window\n"
	"    -nw, --no-window      Run as a dockapp\n"
	"    -s,  --shape          Use XShape extension\n"
	"    -ns, --no-shape       Don't use XShape extension\n"
	"    -c,  --cpunum   <cpu> The cpu to monitor (for SMP systems)\n",
	argv[0]);
 exit(0);
}

void print_version()
{
 printf(WINDOW_NAME " version " VERSION "\n");
 exit(0);
}

void parse_args()
{
 int n;

 for(n = 1; n < argc; n++)
 {
  if(!strcmp(argv[n], "-h") ||
     !strcmp(argv[n], "--help"))
  {
   print_usage();
  }
  else if(!strcmp(argv[n], "-v") ||
          !strcmp(argv[n], "--version"))
  {
   print_version();
  }
  else if(!strcmp(argv[n], "-d") ||
     !strcmp(argv[n], "--display"))
  {
   if(argc <= ++n)
   {
    bad_option(n);
   }
   opt_display = argv[n];
  }
  else if(!strcmp(argv[n], "-m") ||
          !strcmp(argv[n], "--milisecs"))
  {
   if(argc <= ++n)
   {
    bad_option(n);
   }
   opt_milisecs = strtol(argv[n], NULL, 0);
  }
  else if(!strcmp(argv[n], "-b") ||
          !strcmp(argv[n], "--bgcolor"))
  {
   if(argc <= ++n)
   {
    bad_option(n);
   }
   opt_bgcolor = argv[n];
  }
  else if(!strcmp(argv[n], "-f") ||
          !strcmp(argv[n], "--offcolor"))
  {
   if(argc <= ++n)
   {
    bad_option(n);
   }
   opt_offcolor = argv[n];
  }
  else if(!strcmp(argv[n], "-o") ||
          !strcmp(argv[n], "--oncolor"))
  {
   if(argc <= ++n)
   {
    bad_option(n);
   }
   opt_oncolor = argv[n];
  }
  else if(!strcmp(argv[n], "-w") ||
          !strcmp(argv[n], "--window"))
  {
   opt_window = 1;
  }
  else if(!strcmp(argv[n], "-nw") ||
          !strcmp(argv[n], "--no-window"))
  {
   opt_window = 0;
  }
  else if(!strcmp(argv[n], "-s") ||
          !strcmp(argv[n], "--shape"))
  {
   opt_shape = 1;
  }
  else if(!strcmp(argv[n], "-ns") ||
          !strcmp(argv[n], "--no-shape"))
  {
   opt_shape = 0;
  }
  else if(!strcmp(argv[n], "-c") ||
          !strcmp(argv[n], "--cpu"))
  {
   if(argc <= ++n)
   {
    bad_option(n);
   }
   opt_cpunum = strtol(argv[n], NULL, 0);
  }
  else
  {
   fprintf(stderr, "Bad option. Try `%s --help' for more information.\n", argv[0]);
   exit(1);
  }
 }
}

#include "common.c"

int old_cpu_graph = 0;
void draw_window()
{
 int cpu_graph;

 XDrawRectangle(display, buffer, offgc, 19, 5, 39, 10);
 XFillRectangle(display, buffer, backgc, 20, 6, 38, 9);
 if(cpu_total > 0)
  cpu_graph = cpu_used * 40 / cpu_total;
 else cpu_graph = 0;
 if(cpu_graph > 0)
  XDrawRectangle(display, buffer, ongc, 19, 5, cpu_graph - 1, 10);
 if(cpu_graph > 2)
  XDrawRectangle(display, buffer, ongc, 20, 6, cpu_graph - 3, 8);
 if(cpu_graph > 8)
  XFillRectangle(display, buffer, offgc, 23, 9, cpu_graph - 8, 3);
 XCopyArea(display, buffer, buffer, backgc,
           19, 18, 40, 38, 19, 21);
 XFillRectangle(display, buffer, offgc, 19, 18, 40, 2);
 if(old_cpu_graph > 0)
  XFillRectangle(display, buffer, ongc, 19, 18, old_cpu_graph, 2);
 old_cpu_graph = cpu_graph;
}

void proc()
{
 int i;
 fd_set fs;
 int fd = ConnectionNumber(display);

 process_events();
 FD_ZERO(&fs); FD_SET(fd, &fs);
 i = select(fd + 1, &fs, 0, 0, &tv);
 if(i == -1)
 {
  fprintf(stderr, "Error with select(): %s", strerror(errno));
  exit(1);
 }
 if(!i)
 {
  cpu_getusage();
  draw_window();
  update_window();
  tv.tv_sec = opt_milisecs / 1000;
  tv.tv_usec = (opt_milisecs % 1000) * 1000;
 }
}

void free_stuff()
{
 XUnmapWindow(display, mapwindow);
 XDestroyWindow(display, iconwindow);
 XDestroyWindow(display, window);
 XFreePixmap(display, buffer);
 XFreePixmap(display, iconpixmap);
 XFreePixmap(display, pixmask);
 XFreePixmap(display, pixmap);
 XFreeGC(display, ongc);
 XFreeGC(display, offgc);
 XFreeGC(display, backgc);
 XCloseDisplay(display);
}

void set_refresh(menuitem_t *i)
{
 opt_milisecs = i->i;
 
 for(i = m->first; i && i->i > 0; i = i->next)
  if(i->i == opt_milisecs) i->checked = 1;
  else i->checked = 0;

 tv.tv_sec = 0;
 tv.tv_usec = 0;
}

int main(int ac, char **av)
{
 menuitem_t *i;

 argc = ac;
 argv = av;
 parse_args();
 make_window();
 menu_init(display);
 m = menu_new();
 i = menu_append(m, "Refresh: 250ms");
 i->i = 250; i->callback = set_refresh; if(opt_milisecs == 250) i->checked = 1; else i->checked = 0;
 i = menu_append(m, "Refresh: 500ms");
 i->i = 500; i->callback = set_refresh; if(opt_milisecs == 500) i->checked = 1; else i->checked = 0;
 i = menu_append(m, "Refresh: 1s");
 i->i = 1000; i->callback = set_refresh; if(opt_milisecs == 1000) i->checked = 1; else i->checked = 0;
 i = menu_append(m, "Refresh: 2s");
 i->i = 2000; i->callback = set_refresh; if(opt_milisecs == 2000) i->checked = 1; else i->checked = 0;
 i = menu_append(m, "Refresh: 4s");
 i->i = 4000; i->callback = set_refresh; if(opt_milisecs == 4000) i->checked = 1; else i->checked = 0;
 menu_append(m, "Exit");
 cpu_init(opt_cpunum);
 while(!exitloop)
  proc();
 free_stuff();
 return 0;
}
